lib/puppet/configurer.rb

```diff
@@ -360,14 +354,31 @@ def run_internal(options)
         tries += 1
       end
 
-      execute_prerun_command or return nil
+      # now that environment has converged, convert resource catalog into ral catalog
+      # unless we were given a RAL catalog
+      if options[:catalog]
+        ral_catalog = options[:catalog]
+      else
+        # REMIND @duration is the time spent loading the last catalog, and doesn't
+        # account for things like we failed to download and fell back to the cache
+        ral_catalog = convert_catalog(catalog, @duration, options)
+
+        # If not noop, commit the cached resource catalog (not ral catalog). Ideally
+        # we'd just copy the downloaded response body, instead of serializing the
+        # in-memory catalog, but that's hard due to the indirector.
+        indirection = Puppet::Resource::Catalog.indirection
+        if !Puppet[:noop] && indirection.cache?
+          request = indirection.request(:save, nil, catalog, environment: Puppet::Node::Environment.remote(catalog.environment))
+          indirection.cache.save(request)
```

Comment:
I can't find where we were previously updating the cached catalog before this change. I assume that behavior was governed by `options[:ignore_cache_save]`, which I see gets checked in the `save` method in `indirection.rb`.  I think I might be stumbling on indirector magic, which is why I'm having issues tracing this

Reply:
Yep exactly, it's done in https://github.com/puppetlabs/puppet/blob/6.11.0/lib/puppet/indirector/indirection.rb#L226-L234. So we were overwriting the cache, even in cases where the compiled catalog was for the wrong environment.

Reply:
Yep it's buried in https://github.com/puppetlabs/puppet/blob/d984877b2674c89407dedf47307475e22b943e60/lib/puppet/indirector/indirection.rb#L226-L227

Reply:
But where is indirection#find getting called? Is it as a side effect of `convert_catalog`?

Reply:
Ah, the Configurer calls `prepare_and_retrieve_catalog` which calls `retrieve_catalog` which then either tries the cache or not. The latter case calls `retrieve_new_catalog` which calls `Puppet::Resource::Catalog.indirection.find`.

Reply:
Oh! Thank you. I didn't realize that was happening as a part of `prepare_and_retrieve_catalog`